atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/ScalingSweepTaskScheduler.java

```diff
@@ -0,0 +1,176 @@
+                executorService, delay, COOL_DOWN, task, scalingEnabled);
+        scheduler.start(initialThreads);
+        return scheduler;
+    }
+
+    void start(int initialThreads) {
+        for (int i = 0; i < initialThreads; i++) {
+            increaseNumberOfTasks(INITIAL_DELAY);
+        }
+    }
+
+    private synchronized void maybeIncreaseNumberOfTasks(long pause) {
+        lastIncreaseAttempted = Instant.now();
+        if (cooldownPassed(lastModification)) {
+            increaseNumberOfTasks(pause);
+        }
+        scheduleAfterDelay(pause);
+    }
+
+    private synchronized void maybeDecreaseNumberOfTasks(long pause) {
+        if (cooldownPassed(lastModification) && cooldownPassed(lastIncreaseAttempted)) {
+            decreaseNumberOfTasksOrRescheduleIfLast(pause);
+        } else {
+            scheduleAfterDelay(pause);
+        }
+    }
+
+    private synchronized void increaseNumberOfTasks(long pause) {
+        if (runningTasks < 128) {
+            runningTasks++;
+            lastModification = Instant.now();
+            scheduleAfterDelay(pause);
+        }
+    }
+
+    private synchronized void decreaseNumberOfTasksOrRescheduleIfLast(long pause) {
+        if (runningTasks == 1) {
+            scheduleAfterDelay(pause);
+        } else {
+            decreaseNumberOfTasks();
+        }
+    }
+
+    private synchronized void decreaseNumberOfTasks() {
+        runningTasks--;
+        lastModification = Instant.now();
+    }
+
+    private void scheduleAfterDelay(long pause) {
+        executorService.schedule(() -> retryingTask(singleIteration), pause, TimeUnit.MILLISECONDS);
+    }
+
+    private void retryingTask(Callable<SweepIterationResult> task) {
+        try {
+            SweepIterationResult sweepResult = task.call();
+            if (!scalingEnabled.getAsBoolean()) {
+                scheduleAfterDelay(delay.getInitialPause());
+            } else {
+                long pause = delay.getNextPause(sweepResult);
+                SweepIterationResults.caseOf(sweepResult)
+                        .success(numThreads -> determineAction(numThreads, pause))
+                        .unableToAcquireShard(wrap(() -> decreaseNumberOfTasksOrRescheduleIfLast(pause)))
+                        .insufficientConsistency(wrap(() -> scheduleAfterDelay(pause)))
+                        .otherError(wrap(() -> scheduleAfterDelay(pause)))
+                        .disabled(wrap(() -> decreaseNumberOfTasksOrRescheduleIfLast(pause)));
+            }
+        } catch (Exception e) {
+            scheduleAfterDelay(delay.getMaxPause());
+        }
+    }
+
+    private Void determineAction(long numThreads, long pause) {
+        if (numThreads <= BATCH_CELLS_LOW_THRESHOLD) {
+            maybeDecreaseNumberOfTasks(pause);
+        } else if (numThreads >= BATCH_CELLS_HIGH_THRESHOLD) {
+            maybeIncreaseNumberOfTasks(pause);
+        } else {
+            scheduleAfterDelay(pause);
+        }
+        return null;
```

Comment:
I would throw a `SafeIllegalStateException` in this case (unless there's some vagary of the type system I'm missing).